### DISCUSSION TITLE
How to update the field date when findOneAndUpdate

### DISCUSSION THREAD

**Discussion by ZeusZoe** [Automattic/mongoose#14276] [General]
Hi, I've this schema and route:

```
const date=(v)=>{
    console.log(typeof v,v);
    v=new Date();
    return v;
};

const testSchema=new Schema({
    _id:Number,
    name:String,
    createdAt:{type:Date,default:Date.now,immutable:true},
    updatedAt:{type:Date,default:Date.now,set:date}
},{versionKey:false});

router.put("/",async(req,res)=>{
    const c=req.body;
    const d=await testModel.findOneAndUpdate({_id:c._id,updatedAt:c.updatedAt},c);
    res.json(d);
});
```

I noted that the set function (date) is call two onces (why) and the record is not updated when the filter contain 'updatedAt' field too.

I use this data for verify:
```
{
	"_id" : 1,
	"name" : "John Doe",
	"createdAt" : "2024-01-21T08:51:13.826+01:00",
	"updatedAt" : "2024-01-21T10:11:20.422+01:00"
}
```

If I remove the setter function from schema, it work fine.

**Comment by vkarpov15** [comment]
The setter is called twice because Mongoose calls setters on filter properties and update properties. So the setter runs once for `updatedAt` in filter and once for `updatedAt` in update. So `updatedAt` in the filter is updated to the current date, and you get an unexpected query.

In general, I would recommend you use defaults if you're looking to set a default value, not setters. You should also take a look at [Mongoose's timestamps docs](https://mongoosejs.com/docs/timestamps.html) so you don't have to implement timestamps yourself.

**Reply by ZeusZoe** [reply]
Thanks.
Timestamps with findOneAndReplace is not best solution, I believe.
I use findOneAndReplace because my schema include more sub-schema that could be null or default.
I not do want serialize null or empty fields when I need to update the record
